### PR TITLE
P/brent/next obd2 can

### PR DESCRIPTION
**Pull request by brentpicasso**
UI for next gen CAN and OBDII mappings, plus API and related boilerplate. 

Best to test with 2.11.0 firmware (MK3, MK2 with 2.11.0 RC1)

### CODE REVIEW

**Comment by ryandoherty** [comment]
Questions:

1. If I select a preset OBDII channel, then change the tab to OBDII pid and change the pid, the channel name stays on the preselected name. Might be best to change it to 'Uknown' if the user changes any of the other values after selecting a preconfigured OBDII channel.
2. (see image below) Tab ordering sometimes reverses?
image
3. Would be easier to type a number instead of scrolling through hundreds of PID numbers:
image

File: autosportlabs/racecapture/config/rcpconfig.py [hunk 1]
```diff
@@ -828,6 +828,60 @@ def toJson(self):
...
 
+class CANMapping(object):
+    TYPE_UNSIGNED = 0
+    TYPE_SIGNED = 1
+    TYPE_FLOAT = 2
+    ID_MASK_DISABLED = 0
+    CONVERSION_FILTER_DISABLED = 0
+
+    def __init__(self, **kwargs):
+        super(CANMapping, self).__init__(**kwargs)
```
Comment (brentpicasso) [inline autosportlabs/racecapture/config/rcpconfig.py]: ...I don't think so. I think it was just habit that I included it this way from all of the kivy ui widgets.

File: autosportlabs/racecapture/views/configuration/rcp/canchannelsview.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+        self.ids.can_id.text = '{}'.format(can_mapping.can_id)
+
+        self.ids.can_offset_len.text = u'{}({})'.format(can_mapping.offset, can_mapping.length)
+
+        sign = '-' if can_mapping.adder < 0 else '+'
+        self.ids.can_formula.text = u'\u00D7 {} \u00F7 {} {} {}'.format(can_mapping.multiplier, can_mapping.divider, sign, abs(can_mapping.adder))
+        self._loaded = True
+
+class CANPresetResourceCache(ResourceCache):
+    preset_url = "http://podium.live/api/v1/can_presets"
```
Comment (brentpicasso) [inline autosportlabs/racecapture/views/configuration/rcp/canchannelsview.py]: fixing!  funny, resourcemanager isn't fully wired up. probably better have it as a separate PR anyway.

**Comment by brentpicasso** [comment]
It's a balance on cluttering a view with too much information and separating things out into a logical but related sections. the CAN bus screen currently has the baud rate selectors, but needs to be updated to also include the CAN termination switches as well, so it's going to have quite a bit of information on one screen already.

A lot of this will go to automated setup wizards when the app is first installed where a series of questions and automated setup will guide people through things.

**Comment by brentpicasso** [comment]
Regarding pre-set selection.  Actually, what's being attempted here is if they select an existing SAE standard channel, the rest of the information is automatically populated throughout the rest of the dialog. 

However, there's a distinct case where they may want to add an OEM specific PID. They're going to set the channel name and discretely customize the PID, formula and etc.  I don't think we can symmetrically go and set the channel name to 'unknown' if they change the PID (for example) - because then they might go change the channel name back again, and then it might overwrite their PID changes - a catch 22.

Reflecting on it more, I think we need to pop up a warning or ask the user if they want the selected preset to change the existing settings.  But maybe only if they're editing a current OBDII channel . If they're adding a new channel, then the preset populates the channel settings quietly.

**Comment by brentpicasso** [comment]
I debated the scrolling through PIDs vs typing a number. I do like how it's crisply constrained to the numbers we specify. I think it's pretty quick to scroll through overall: 1 - 255.  I'd like to keep it unless it proves to be cumbersome based on feedback.

File: autosportlabs/racecapture/config/rcpconfig.py (hunk 1 above)
Comment (brentpicasso) [inline autosportlabs/racecapture/config/rcpconfig.py]: fixed.

**Comment by brentpicasso** [comment]
ok @ryandoherty  - I made a change to how presets are applied. 

* if the channel editor is for a brand new channel, we never prompt. 
* if the current channel matches the incoming preset exactly, then we don't prompt to overwrite preset
* If editing an existing channel *and* the channel was previously customized (and no longer matches the preset for the channel matching that name),  we prompt the user if they want to apply the preset- possibly avoiding the possible overwriting of the channel mapping details.

**Comment by brentpicasso** [comment]
Fixed tab ordering, made preset selection more intelligent, avoiding cases where we overwrite user's customizations. 

Added help popups, too.

**Comment by brentpicasso** [comment]
resolves #984 #982

**Comment by ryandoherty** [comment]
Awesome! The tabs are really nice, good work on those 👍

**Pull request merged**